File: app/src/lib/native/shared-app-state.ts
import type { SemanticState } from '@/constants/state-tokens';

/**
 * Shape of the JSON blob that the RN app writes into the App Group
 * shared container (`group.app.releasepilot.shared`) at key
 * `release-pilot.state.v1`.
 *
 * Both the WidgetKit widget AND the Live Activity views deserialize
 * THIS exact shape on the Swift side (see `targets/widget/SharedAppState.swift`).
 *
 * IMPORTANT: If you add/rename a field here, you MUST bump the version
 * envelope (`v`) and update the Swift decoder. Mismatch = stale or
 * broken widget.
 *
 * Schema versioning policy:
 *   The `v` envelope is reserved for BREAKING shape changes. Field
 *   additions are made WITHOUT bumping `v`, because Swift's Codable
 *   ignores unknown fields by default — so old widget binaries
 *   gracefully ignore new fields, and new widget binaries can read old
 *   payloads because the new fields are declared Optional on the
 *   Swift side.
 *
 *   v1 — initial release: { v, lastUpdatedMs, apps }
 *        — additive (no bump): proStatus, headline
 */

export type WidgetAppRow = {
  /** ASC app id — stable across runs. */
  ascId: string;
  /** App's display name (e.g. "Recall: Personal Memory"). */
  name: string;
  /** Bundle identifier (e.g. "com.acme.recall"). */
  bundleId: string;
  /** Our 7 semantic states. */
  state: SemanticState;
  /** Human-readable label e.g. "Live on App Store". */
  stateLabel: string;
  /** Compact label e.g. "Live". */
  stateShortLabel: string;
  /** Version e.g. "1.8.23". */
  versionString: string;
  /** Build number e.g. "29" — nullable when no build attached. */
  buildNumber: string | null;
  /** ISO 8601 timestamp of when this state was last observed. */
  lastChangedAt: string;
  /** Hex string e.g. "#7A5C00" — foreground tint for the state badge. */
  stateFgLight: string;
  stateFgDark: string;
  /** Background tint for the state badge. */
  stateBgLight: string;
  stateBgDark: string;
};

/**
 * Subscription state visible to the widget. The widget renders
 * differently for each:
 *  - `pro`     → full data (up to N apps), no CTA banner
 *  - `free`    → 1 app + soft "Upgrade for all apps" headline
 *  - `lapsed`  → 1 app + prominent "Renew Pro" headline (user used to
 *                be Pro within the last 60 days; we want to nudge them
 *                back without nagging forever)
 */
export type WidgetProStatus = 'pro' | 'free' | 'lapsed';

export type SharedAppState = {
  v: 1;
  /** Epoch milliseconds when the RN app last wrote this state. */
  lastUpdatedMs: number;
  /** Apps sorted by `HERO_PRIORITY` ascending (most-urgent first), so
   *  `apps[0]` is the right pick for the lock-screen rectangle + home
   *  small widget. Cap depends on `proStatus` (free/lapsed = 1). */
  apps: WidgetAppRow[];
  /** Subscription state (drives headline + apps cap). Additive field —
   *  old Swift binaries (pre-tier-aware widget) silently ignore. */
  proStatus: WidgetProStatus;
  /** Optional banner copy. When present, the widget renders this
   *  prominently with a "tap to open Release Pilot" affordance.
   *  Additive field — old Swift binaries silently ignore. */
  headline: string | null;
};

/** Key under UserDefaults(suiteName:) used by both the writer and Swift. */
export const SHARED_STATE_KEY = 'release-pilot.state.v1';

/** App Group suite name (matches `entitlements` in `app.json`). */
export const APP_GROUP_ID = 'group.app.releasepilot.shared';

/** How many apps the widget shows for each tier. Pro is effectively
 *  unbounded (Apple caps a `systemLarge` at ~6 visible rows anyway). */
export const WIDGET_APPS_CAP: Record<WidgetProStatus, number> = {
  pro: 99,
  free: 1,
  lapsed: 1,
};

File: app/targets/widget/SharedAppState.swift
import Foundation
import SwiftUI

/// Mirror of the TypeScript `SharedAppState` shape defined in
/// `src/lib/native/shared-app-state.ts`.
///
/// IMPORTANT: any change to fields MUST be made on both sides.
/// The `v` envelope number lets the Swift decoder bail on mismatched
/// versions rather than silently rendering stale shapes.
struct WidgetAppRow: Codable, Identifiable, Hashable {
    let ascId: String
    let name: String
    let bundleId: String
    let state: String           // semantic state — matches TS `SemanticState`
    let stateLabel: String
    let stateShortLabel: String
    let versionString: String
    let buildNumber: String?
    let lastChangedAt: String   // ISO 8601
    let stateFgLight: String    // hex e.g. "#7A5C00"
    let stateBgLight: String
    let stateFgDark: String
    let stateBgDark: String

    var id: String { ascId }
}

/// Subscription tier visible to the widget — drives apps cap +
/// "Renew Pro" headline. Must match TS `WidgetProStatus`.
///
/// Optional in the payload — old TS bundles (pre-tier-aware widget)
/// don't write this field. When missing, the widget defaults to `.pro`
/// (i.e. show all apps, no headline) so existing users don't lose data
/// during a JS-bundle-only upgrade.
enum WidgetProStatus: String, Codable {
    case pro
    case free
    case lapsed
}

struct SharedAppState: Codable {
    let v: Int
    let lastUpdatedMs: Double
    let apps: [WidgetAppRow]
    /// Additive field — Optional so old payloads (without it) still
    /// decode successfully. Defaults to `.pro` at render time.
    let proStatus: WidgetProStatus?
    /// Optional CTA banner (e.g. "Renew Pro to track all apps"). When
    /// nil the widget renders without any banner.
    let headline: String?
}

/// Loads the latest snapshot from the App Group container.
/// Returns `nil` on first launch (before the RN app has ever written) OR
/// on schema-version mismatch — both treated as "no data, show empty state".
enum SharedAppStateLoader {
    static let appGroupId = "group.app.releasepilot.shared"
    static let stateKey   = "release-pilot.state.v1"
    /// Matches `SharedAppState.v` in `src/lib/native/shared-app-state.ts`.
    /// Reserved for BREAKING shape changes — field additions are done
    /// without bumping (Codable ignores unknown fields, Optional fields
    /// gracefully fall back to nil on read).
    static let schemaV    = 1

    static func load() -> SharedAppState? {
        guard let defaults = UserDefaults(suiteName: appGroupId) else { return nil }
        guard let data = defaults.string(forKey: stateKey)?.data(using: .utf8) else { return nil }
        do {
            let decoded = try JSONDecoder().decode(SharedAppState.self, from: data)
            guard decoded.v == schemaV else { return nil }
            return decoded
        } catch {
            return nil
        }
    }
}

// MARK: - Color helpers

/// Build a SwiftUI Color from a "#RRGGBB" string. Falls back to gray on
/// malformed input — better than crashing the widget render.
extension Color {
    init(hex: String) {
        let trimmed = hex.trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "#", with: "")
        var rgb: UInt64 = 0
        guard Scanner(string: trimmed).scanHexInt64(&rgb), trimmed.count == 6 else {
            self = .gray
            return
        }
        let r = Double((rgb & 0xFF0000) >> 16) / 255.0
        let g = Double((rgb & 0x00FF00) >>  8) / 255.0
        let b = Double( rgb & 0x0000FF       ) / 255.0
        self = Color(red: r, green: g, blue: b)
    }
}

/// Picks the right hex for the current color scheme.
struct AdaptiveHexColor {
    let light: String
    let dark:  String

    func resolve(_ scheme: ColorScheme) -> Color {
        Color(hex: scheme == .dark ? dark : light)
    }
}

extension WidgetAppRow {
    var fg: AdaptiveHexColor { AdaptiveHexColor(light: stateFgLight, dark: stateFgDark) }
    var bg: AdaptiveHexColor { AdaptiveHexColor(light: stateBgLight, dark: stateBgDark) }
}
